File: Programmers/행렬 테두리 회전하기.py
def solution(rows, columns, queries):
    answer = []

    matrix = []
    for i in range(1, rows + 1):
        temp = []
        for k in range(1, columns + 1):
            temp.append((i - 1) * columns + k)
        matrix.append(temp)
    for query in queries:
        answer.append(rotate(query, matrix))
    rotate(queries[0], matrix)

    return answer


def rotate(query, matrix):
    move_number = []
    # 왼쪽 세로부터
    move_number.append(matrix[query[0] - 1][query[1] - 1])
    for i in range(query[2] - query[0]):
        temp = matrix[query[0] - 1 + i][query[1] - 1]
        move_number.append(matrix[query[0] + i][query[1] - 1])
        matrix[query[0] - 1 + i][query[1] - 1] = matrix[query[0] + i][query[1] - 1]
        matrix[query[0] + i][query[1] - 1] = temp

    # 아래 가로
    for i in range(query[3] - query[1]):
        temp = matrix[query[2] - 1][query[1] - 1 + i]
        move_number.append(matrix[query[2] - 1][query[1] + i])
        matrix[query[2] - 1][query[1] - 1 + i] = matrix[query[2] - 1][query[1] + i]
        matrix[query[2] - 1][query[1] + i] = temp

    # 오른쪽 세로
    for i in range(query[2] - query[0]):
        temp = matrix[query[2] - 1 - i][query[3] - 1]
        move_number.append(matrix[query[2] - 2 - i][query[3] - 1])
        matrix[query[2] - 1 - i][query[3] - 1] = matrix[query[2] - 2 - i][query[3] - 1]
        matrix[query[2] - 2 - i][query[3] - 1] = temp

    # 위 가로
    for i in range(query[3] - query[1] - 1):
        temp = matrix[query[0] - 1][query[3] - 1 - i]
        move_number.append(matrix[query[0] - 1][query[3] - 2 - i])
        matrix[query[0] - 1][query[3] - 1 - i] = matrix[query[0] - 1][query[3] - 2 - i]
        matrix[query[0] - 1][query[3] - 2 - i] = temp

    move_number = set(move_number)
    return min(move_number)

File: Programmers JS/행렬 테두리 회전하기.js
function solution(rows, columns, queries) {
  let answer = [];
  let board = [];
  for (let i = 1; i < rows + 1; i++) {
    let temp = [];
    for (let j = 1; j < columns + 1; j++) {
      temp.push((i - 1) * columns + j);
    }
    board.push(temp);
  }

  for (let query of queries) {
    let numSet = [];
    board = rotate(query, board, numSet);
    answer.push(Math.min(...numSet));
  }

  return answer;
}

function rotate(query, board, numSet) {
  let start_row = query[0] - 1;
  let start_col = query[1] - 1;
  let finish_row = query[2] - 1;
  let finish_col = query[3] - 1;
  numSet.push(board[start_row][start_col]);
  //왼쪽 세로
  for (let i = start_row; i < finish_row; i++) {
    let temp = board[i][start_col];

    board[i][start_col] = board[i + 1][start_col];
    numSet.push(board[i][start_col]);
    board[i + 1][start_col] = temp;
  }
  //아래 가로
  for (let i = start_col; i < finish_col; i++) {
    let temp = board[finish_row][i];

    board[finish_row][i] = board[finish_row][i + 1];
    numSet.push(board[finish_row][i]);
    board[finish_row][i + 1] = temp;
  }
  //오른쪽 세로
  for (let i = finish_row; i > start_row; i--) {
    let temp = board[i][finish_col];
    board[i][finish_col] = board[i - 1][finish_col];
    numSet.push(board[i][finish_col]);
    board[i - 1][finish_col] = temp;
  }
  // 위쪽 가로
  for (let i = finish_col; i > start_col + 1; i--) {
    let temp = board[start_row][i];
    board[start_row][i] = board[start_row][i - 1];
    numSet.push(board[start_row][i]);
    board[start_row][i - 1] = temp;
  }

  return board;
}
